File: org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/form/fields/groupbox/internal/VerticalSmartGroupBoxBodyGrid.java
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.rt.client.ui.form.fields.groupbox.internal;

import static org.eclipse.scout.rt.platform.util.NumberUtility.divideAndCeil;

import java.util.List;

import org.eclipse.scout.rt.client.ui.form.fields.GridData;
import org.eclipse.scout.rt.client.ui.form.fields.IFormField;
import org.eclipse.scout.rt.client.ui.form.fields.groupbox.internal.matrix.VerticalGridMatrix;

/**
 * @author Andreas Hoegger
 * @since 4.0.0 M6 25.02.2014
 */
public class VerticalSmartGroupBoxBodyGrid extends AbstractGroupBoxBodyGrid {

  @Override
  protected void layoutAllDynamic(List<IFormField> fields) {
    // calculate the used cells
    int cellCount = 0;
    for (IFormField f : fields) {
      GridData hints = getGridDataFromHints(f, getGridColumnCount());
      cellCount += hints.w * hints.h;
    }

    int rowCount = divideAndCeil(cellCount, getGridColumnCount());
    VerticalGridMatrix matrix = new VerticalGridMatrix(getGridColumnCount(), rowCount);
    while (!matrix.computeGridData(fields)) {
      matrix.resetAll(getGridColumnCount(), ++rowCount);
    }
    // setGridData
    for (IFormField f : fields) {
      GridData data = matrix.getGridData(f);
      f.setGridDataInternal(data);
    }
    setGridRows(matrix.getRowCount());
  }
}

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/form/fields/groupbox/grid/VerticalSmartGroupBoxBodyGrid.js
/*******************************************************************************
 * Copyright (c) 2014-2015 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 ******************************************************************************/
scout.VerticalSmartGroupBoxBodyGrid = function() {
  scout.VerticalSmartGroupBoxBodyGrid.parent.call(this);
};
scout.inherits(scout.VerticalSmartGroupBoxBodyGrid, scout.GroupBoxBodyGrid);

scout.VerticalSmartGroupBoxBodyGrid.prototype.layoutAllDynamic = function(fields) {
  var cellCount = 0;
  fields.forEach(function(f) {
    var hints = scout.GroupBoxBodyGrid.getGridDataFromHints(f, this.getGridColumnCount());
    cellCount += hints.w * hints.h;
  }.bind(this));

  // do the calc
  var rowCount = Math.floor((cellCount + this.getGridColumnCount() - 1) / this.getGridColumnCount());
  var matrix = new scout.VerticalGridMatrixGroupBox(this.getGridColumnCount(), rowCount);
  while (!matrix.computeGridData(fields)) {
    matrix.resetAll(this.getGridColumnCount(), ++rowCount);
  }

  // set gridData
  fields.forEach(function(f) {
    f.gridData = matrix.getGridData(f);
  });
  this.gridRows = matrix.getRowCount();
};
